Mathlib/Data/Nat/Init.lean

```diff
@@ -434,6 +434,45 @@ protected lemma div_left_inj (hda : d ∣ a) (hdb : d ∣ b) : a / d = b / d ↔
   refine ⟨fun h ↦ ?_, congrArg fun b ↦ b / d⟩
   rw [← Nat.mul_div_cancel' hda, ← Nat.mul_div_cancel' hdb, h]
 
+/-- TODO: Replace `Nat.sub_mul_div` in core. -/
+lemma sub_mul_div' (a b c : ℕ) : (a - b * c) / b = a / b - c := by
+  obtain h | h := Nat.le_total (b * c) a
+  · rw [Nat.sub_mul_div _ _ _ h]
+  · rw [Nat.sub_eq_zero_of_le h, Nat.zero_div]
+    by_cases hn : b = 0
+    · simp only [hn, Nat.div_zero, zero_le, Nat.sub_eq_zero_of_le]
+    · have h2 : a / b ≤ (b * c) / b := Nat.div_le_div_right h
+      rw [Nat.mul_div_cancel_left _ (zero_lt_of_ne_zero hn)] at h2
+      rw [Nat.sub_eq_zero_of_le h2]
+
+lemma mul_sub_div_of_dvd (hc : c ≠ 0) (hcb : c ∣ b) (a : ℕ) : (c * a - b) / c = a - b / c := by
+  obtain ⟨_, hx⟩ := hcb
+  simp only [hx, ← Nat.mul_sub_left_distrib, Nat.mul_div_right, zero_lt_of_ne_zero hc]
+
+lemma mul_add_mul_div_of_dvd (hb : b ≠ 0) (hd : d ≠ 0) (hba : b ∣ a) (hdc : d ∣ c) :
+    (a * d + b * c) / (b * d) = a / b + c / d := by
+  obtain ⟨n, hn⟩ := hba
+  obtain ⟨_, hm⟩ := hdc
+  rw [hn, hm, Nat.mul_assoc b n d, Nat.mul_comm n d, ← Nat.mul_assoc, ← Nat.mul_assoc,
+    ← Nat.mul_add, Nat.mul_div_right _ (zero_lt_of_ne_zero hb),
+    Nat.mul_div_right _ (zero_lt_of_ne_zero hd),
+    Nat.mul_div_right _ (Nat.mul_pos (zero_lt_of_ne_zero hb) (zero_lt_of_ne_zero hd))]
+
+lemma mul_sub_mul_div_of_dvd (hb : b ≠ 0) (hd : d ≠ 0) (hba : b ∣ a) (hdc : d ∣ c) :
+    (a * d - b * c) / (b * d)  = a / b - c / d:= by
+  obtain ⟨n, hn⟩ := hba
+  obtain ⟨m, hm⟩ := hdc
+  rw [hn, hm]
+  rw [Nat.mul_assoc,Nat.mul_comm n d, ← Nat.mul_assoc,← Nat.mul_assoc, ← Nat.mul_sub_left_distrib,
+    Nat.mul_div_right _ (zero_lt_of_ne_zero hb), Nat.mul_div_right _ (zero_lt_of_ne_zero hd),
+    Nat.mul_div_right _ (Nat.mul_pos (zero_lt_of_ne_zero hb) (zero_lt_of_ne_zero hd))]
+
+lemma div_mul_assoc (hba : b ∣ a) (c : ℕ) : (a / b) * c = (a * c) / b := by
```

Comment:
```suggestion
protected lemma div_mul_right_comm (hba : b ∣ a) (c : ℕ) : (a / b) * c = (a * c) / b := by
```
since this lemma right-commutes multiplication and division, and since there could be a corresponding lemma in the root namespace. Also the brackets are unnecessary
```suggestion
protected lemma div_mul_right_comm (hba : b ∣ a) (c : ℕ) : a / b * c = a * c / b := by
```
Could you please also add `Nat.mul_div_right_comm` (which is just `(Nat.div_mul_right_comm ..).symm`)?

Reply:
Sure.

One questtion: why the `protected` tag?

Reply:
Because of:
> since there could be a corresponding lemma in the root namespace

`mul_div_right_comm` already exists as a lemma, and you don't want Lean to be confused between `mul_div_right_comm` and `Nat.mul_div_right_comm` after you've opened the `Nat` namespace (`protected lemma Foo.bar` forbids `bar` from meaning `Foo.bar` after an `open Foo`). `div_mul_right_comm` doesn't exist yet, but it seems fitting to protect `Nat.div_mul_right_comm` just in case

Reply:
Thanks!